core/shared_obj.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2017, Nefeli Networks, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// * Neither the names of the copyright holders nor the names of their
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_SHARED_OBJ_H_
+#define BESS_SHARED_OBJ_H_
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+
+#include <glog/logging.h>
+
+namespace bess {
+
+using SharedObjectKey = std::pair<std::type_index, std::string>;
+
+struct PairHasher {
+  template <typename T1, typename T2>
+  std::size_t operator()(const std::pair<T1, T2> &p) const noexcept {
+    return std::hash<T1>{}(p.first) + std::hash<T2>{}(p.second);
+  }
+};
+
+extern std::unordered_map<SharedObjectKey, std::weak_ptr<void>, PairHasher>
+    shared_objects;
+
+template <typename T>
+std::shared_ptr<T> GetSharedObject(const std::string &name,
+                                   bool lookup_only = false) {
```

Comment:
Might it be useful to allow the caller to pass an already-constructed `T*` or arguments for construction of a `T`?  (Since `T` may not have a zero-arg constructor.)

Reply:
Ah good point. Probably I would add a `std::function` argument to be used for object construction, rather than directly give it constructor arguments. Preparing arguments or a constructed instance a priori might be expensive/inapplicable in some cases...

Reply:
Sounds good.